src/engine/CMakeLists.txt

```diff
@@ -9,7 +9,7 @@ file(GLOB_RECURSE RENDER_SOURCES render/*.cpp render/*.hpp)
 if(WIN32)
 set (GL_CORE_SRCS ${CMAKE_SOURCE_DIR}/src/engine/render/opengl_renderer/gl_core_4_4.c)
```

Comment:
We need to use 3.3 on all platforms, please remove this branching

Reply:
why ? we can support 4/3 easy

Reply:
don't need to

Reply:
with this you create fragmentation without no need

Reply:
with this change we will limit users from getting new openGL features

Reply:
OpenGL major version numbers effectively represent hardware levels, like D3D major version numbers. GL 2.x represents DX9 hardware. GL 3.x represents DX10-class hardware. And therefore GL 4.x represents DX11-class hardware.

Not all features introduced in GL 4.x are hardware-specific, however. Some are API convenience and cleanup features. These are available as extensions to lower-level hardware, and drivers from NVIDIA and AMD do provide them. The biggest of these is program separation, which is something GLSL has needed for a long time.

In terms of actual hardware features presented by GL 4.x that are not available in 3.x, it isn't huge or game-changing. It's certainly significant, but not earth-moving. You can find a detailed breakdown of GL features by version on the OpenGL Wiki.

Image load/store and atomic counters are probably the biggest ones (this includes shader storage buffers too), though simultaneously the least advertised. Shaders get to write/read from images and buffers directly, though there are a lot of caveats and synchronization issues to understand about doing this.

GL 4.3 added compute shaders, which coupled with the above allow you to do computational tasks without the pain of OpenCL/OpenGL interop.

The feature that gets talked about the most is tessellation. This introduces an additional pair of shader stages: one that decides how much to tessellate a primitive, and the other to decide how to generate the new values from the tessellated primitive.

Subroutines are an interesting but often overlooked feature. Basically, you can attach subroutines to shader programs like uniforms. So you could use the same basic program that will run a lighting function, but you can swap lighting functions without changing the program. The basic program would fetch textures, figure out what the diffuse/specular/etc material parameters are, determine the normal and lighting parameters, and pass that to the subroutine. You could have a wide variety of lighting subroutines, from the quick and simple Lambertian diffuse to full-on Cook-Torrance (though you'd need some more material parameters).